From ce97e4adbaa8d6ad1e59c91ed9b844bca8372131 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Mon, 1 Jul 2019 10:37:35 +0800
Subject: [PATCH 06/18] media: capture: linux: Support libv4l2 plugins

Allow using libv4l2 plugins for linux v4l2 capture devices.

Depends on custom libv4l2 with mmap & munmap.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 media/capture/video/linux/BUILD.gn            | 15 ++++++
 .../video/linux/v4l2_capture_device_impl.cc   | 53 ++++++++++++++++++-
 .../video/linux/v4l2_capture_device_impl.h    |  8 +++
 3 files changed, 75 insertions(+), 1 deletion(-)

diff --git a/media/capture/video/linux/BUILD.gn b/media/capture/video/linux/BUILD.gn
index 8f812f723..6ffe7345d 100644
--- a/media/capture/video/linux/BUILD.gn
+++ b/media/capture/video/linux/BUILD.gn
@@ -2,10 +2,19 @@
 # Use of this source code is governed by a BSD-style license that can be
 # found in the LICENSE file.
 
+import("//build/buildflag_header.gni")
 import("//third_party/webrtc/webrtc.gni")
 
 assert(is_linux || is_chromeos)
 
+buildflag_header("buildflags") {
+  header = "buildflags.h"
+
+  flags = [
+    "USE_LIBV4L2=$use_v4lplugin",
+  ]
+}
+
 source_set("linux") {
   visibility = [
     ":test_support",
@@ -70,6 +79,12 @@ source_set("linux") {
     "//ui/display",
     "//ui/gfx",
   ]
+
+  public_deps += [ ":buildflags" ]
+
+  if (use_v4lplugin) {
+    deps += [ "//media/gpu/v4l2:libv4l2_stubs" ]
+  }
 }
 
 source_set("test_support") {
diff --git a/media/capture/video/linux/v4l2_capture_device_impl.cc b/media/capture/video/linux/v4l2_capture_device_impl.cc
index c8a49b3f0..b9fa8efd5 100644
--- a/media/capture/video/linux/v4l2_capture_device_impl.cc
+++ b/media/capture/video/linux/v4l2_capture_device_impl.cc
@@ -10,19 +10,62 @@
 #include <sys/poll.h>
 #include <unistd.h>
 
+#if BUILDFLAG(USE_LIBV4L2)
+// Auto-generated for dlopen libv4l2 libraries
+#include "media/gpu/v4l2/v4l2_stubs.h"
+#include "third_party/v4l-utils/lib/include/libv4l2.h"
+
+using media_gpu_v4l2::kModuleV4l2;
+using media_gpu_v4l2::InitializeStubs;
+using media_gpu_v4l2::StubPathMap;
+
+inline static constexpr char kLibV4l2Path[] =
+#if defined(__aarch64__)
+      "/usr/lib64/libv4l2.so";
+#else
+      "/usr/lib/libv4l2.so";
+#endif
+#endif
+
 namespace media {
 
 V4L2CaptureDeviceImpl::~V4L2CaptureDeviceImpl() = default;
 
+V4L2CaptureDeviceImpl::V4L2CaptureDeviceImpl() {
+#if BUILDFLAG(USE_LIBV4L2)
+  use_libv4l2_ = false;
+#endif
+}
+
 int V4L2CaptureDeviceImpl::open(const char* device_name, int flags) {
-  return ::open(device_name, flags);
+  int fd = ::open64(device_name, flags);
+  if (fd < 0)
+    return fd;
+
+#if BUILDFLAG(USE_LIBV4L2)
+  StubPathMap paths;
+  paths[kModuleV4l2].push_back(kLibV4l2Path);
+
+  static bool libv4l2_initialized = InitializeStubs(paths);
+  if (libv4l2_initialized && v4l2_fd_open(fd, V4L2_DISABLE_CONVERSION) != -1)
+      use_libv4l2_ = true;
+#endif
+  return fd;
 }
 
 int V4L2CaptureDeviceImpl::close(int fd) {
+#if BUILDFLAG(USE_LIBV4L2)
+  if (use_libv4l2_)
+    return v4l2_close(fd);
+#endif
   return ::close(fd);
 }
 
 int V4L2CaptureDeviceImpl::ioctl(int fd, int request, void* argp) {
+#if BUILDFLAG(USE_LIBV4L2)
+  if (use_libv4l2_)
+    return v4l2_ioctl(fd, request, argp);
+#endif
   return ::ioctl(fd, request, argp);
 }
 
@@ -32,10 +75,18 @@ void* V4L2CaptureDeviceImpl::mmap(void* start,
                                   int flags,
                                   int fd,
                                   off_t offset) {
+#if BUILDFLAG(USE_LIBV4L2)
+  if (use_libv4l2_)
+    return v4l2_mmap(start, length, prot, flags, fd, offset);
+#endif
   return ::mmap(start, length, prot, flags, fd, offset);
 }
 
 int V4L2CaptureDeviceImpl::munmap(void* start, size_t length) {
+#if BUILDFLAG(USE_LIBV4L2)
+  if (use_libv4l2_)
+    return v4l2_munmap(start, length);
+#endif
   return ::munmap(start, length);
 }
 
diff --git a/media/capture/video/linux/v4l2_capture_device_impl.h b/media/capture/video/linux/v4l2_capture_device_impl.h
index e2f6bc154..336d1504e 100644
--- a/media/capture/video/linux/v4l2_capture_device_impl.h
+++ b/media/capture/video/linux/v4l2_capture_device_impl.h
@@ -10,6 +10,7 @@
 
 #include "media/capture/capture_export.h"
 #include "media/capture/video/linux/v4l2_capture_device.h"
+#include "media/capture/video/linux/buildflags.h"
 
 namespace media {
 
@@ -17,6 +18,8 @@ namespace media {
 // V4L2 APIs.
 class CAPTURE_EXPORT V4L2CaptureDeviceImpl : public V4L2CaptureDevice {
  public:
+  V4L2CaptureDeviceImpl();
+
   int open(const char* device_name, int flags) override;
   int close(int fd) override;
   int ioctl(int fd, int request, void* argp) override;
@@ -32,6 +35,11 @@ class CAPTURE_EXPORT V4L2CaptureDeviceImpl : public V4L2CaptureDevice {
 
  private:
   ~V4L2CaptureDeviceImpl() override;
+
+#if BUILDFLAG(USE_LIBV4L2)
+  // Use libv4l2 when operating |fd|.
+  bool use_libv4l2_;
+#endif
 };
 
 }  // namespace media
-- 
2.39.5

